### ISSUE TITLE
Distilling labels as well as images

### ISSUE THREAD

**Issue by xu-ji** [SsnL/dataset-distillation#26]
It seems that you fix the label associated with each image when producing a weight update with it.

This, I presume, helps the images you learn be specialised to single classes. E.g. the MNIST digits you produce (for random initialisation) are clearly distinct digits. 

However, the labels are also a vital part of the dataset.

Did you consider randomly initialising them and backpropagating the training signal to learn them too?

**Comment by ilia10000** [comment]
We actually demonstrated this in a recent paper where we extended dataset distillation:
Code: https://github.com/ilia10000/dataset-distillation
Paper: https://arxiv.org/abs/1910.02551v2

**Comment by xu-ji** [comment]
Thank you. It seems that in your work, the learned labels are initialised to hard values.

Can you give some more details about what happened when you tried random initialisation of labels, which in section 3.2 you say does badly? What was the performance?

**Comment by ilia10000** [comment]
In cases where there are a moderate number of images (e.g. 1+ distilled image per class), there is a noticeable drop in performance if soft-labels are initialized completely randomly before being learned instead of initializing as hard labels before being learned. On MNIST, that was about a 1% difference if using 10 distilled images. On CIFAR, I believe it was 2-5% when using 100 distilled images. 

However, in cases where we used much fewer images than classes, there was interesting behavior. Generally, randomly initialization was actually better here (e.g. double-digit difference in performance when using a single distilled image to represent all of MNIST). Hard label initialization could perform better than random initialization only if the hard labels were chosen very carefully (e.g. for representing MNIST with 5 images, we got the best performance by starting with the following label initializations: 1 2 4 5 9).

We think this is because using the hard initializations leads to better class separation early on in the distillation process. It's possible that there are smarter random initializations for the labels than the uniform random one we used. Perhaps it might be more optimal to start with smoothed versions of the hard labels? We haven't explored this.
